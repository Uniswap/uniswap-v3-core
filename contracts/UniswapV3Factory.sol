// SPDX-License-Identifier: GPL-3.0-or-later
pragma solidity =0.6.12;

import './interfaces/IUniswapV3Factory.sol';

import './UniswapV3Pair.sol';

/// @notice The Uniswap V3 Factory.
/// @notice A factory for creating new V3 trading pairs.
/// @dev Creates new pairs at deterministic addresses.
contract UniswapV3Factory is IUniswapV3Factory {
    address public override owner;

    mapping(uint24 => int24) public override feeAmountTickSpacing;
    uint24[] public override allEnabledFeeAmounts;

    mapping(address => mapping(address => mapping(uint24 => address))) public override getPair;
    address[] public override allPairs;

    /// @notice Gets length of allPairs array.
    /// @return The length of allPairs address array.
    function allPairsLength() external view override returns (uint256) {
        return allPairs.length;
    }

    /// @notice Gets length of allEnabledFeeAmounts array.
    /// @return Length of allEnabledFeeAmounts array.
    function allEnabledFeeAmountsLength() external view override returns (uint256) {
        return allEnabledFeeAmounts.length;
    }

    /// @notice The Factory contract constructor.
    /// @param _owner The owner of the Factory contract.
    constructor(address _owner) public {
        owner = _owner;
        emit OwnerChanged(address(0), _owner);

        _enableFeeAmount(600, 1);
        _enableFeeAmount(3000, 1);
        _enableFeeAmount(9000, 1);
    }

    /// @notice Deploys a new trading pair.
    /// @param tokenA the first token of the desired pair.
    /// @param tokenB the second token of the desired pair.
    /// @param fee the desired fee.
    /// @return pair Returns the address of the newly deployed pair.
    function createPair(
        address tokenA,
        address tokenB,
        uint24 fee
    ) external override returns (address pair) {
        require(tokenA != tokenB, 'UniswapV3Factory::createPair: tokenA cannot be the same as tokenB');
        (address token0, address token1) = tokenA < tokenB ? (tokenA, tokenB) : (tokenB, tokenA);
        require(token0 != address(0), 'UniswapV3Factory::createPair: tokens cannot be address 0');
        int24 tickSpacing = feeAmountTickSpacing[fee];
        require(tickSpacing != 0, 'UniswapV3Factory::createPair: fee amount is not allowed');
        require(getPair[token0][token1][fee] == address(0), 'UniswapV3Factory::createPair: pair already exists');
        // CREATE2 salt is 0 since token0, token1, and fee are included as constructor arguments
        pair = address(new UniswapV3Pair{salt: bytes32(0)}(address(this), token0, token1, fee, tickSpacing));
        allPairs.push(pair);
        getPair[token0][token1][fee] = pair;
        // populate mapping in the reverse direction, deliberate choice to avoid the cost of comparing addresses
        getPair[token1][token0][fee] = pair;
        emit PairCreated(token0, token1, fee, tickSpacing, pair, allPairs.length);
    }

    /// @notice Sets Factory contract owner to a new address.
    /// @param _owner The new owner of the factory contract.
    /// @dev only callable by current owner of factory contract.
    function setOwner(address _owner) external override {
        require(msg.sender == owner, 'UniswapV3Factory::setOwner: must be called by owner');
        emit OwnerChanged(owner, _owner);
        owner = _owner;
    }

    /// @dev see enableFeeAmount.
    function _enableFeeAmount(uint24 fee, int24 tickSpacing) private {
        require(fee < 1000000, 'UniswapV3Factory::_enableFeeAmount: fee amount be greater than or equal to 100%');
        require(feeAmountTickSpacing[fee] == 0, 'UniswapV3Factory::_enableFeeAmount: fee amount is already enabled');
        require(tickSpacing > 0, 'UniswapV3Factory::_enableFeeAmount: tick spacing must be greater than 0');

        feeAmountTickSpacing[fee] = tickSpacing;
        allEnabledFeeAmounts.push(fee);
        emit FeeAmountEnabled(fee, tickSpacing);
    }

    /// @notice If chosen, enables the fee option when a pair is deployed.
    /// @param fee The chosen fee option - passed via createPair.
    function enableFeeAmount(uint24 fee, int24 tickSpacing) external override {
        require(msg.sender == owner, 'UniswapV3Factory::enableFeeAmount: must be called by owner');

        _enableFeeAmount(fee, tickSpacing);
    }
}
