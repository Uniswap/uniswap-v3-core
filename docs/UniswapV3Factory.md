## `UniswapV3Factory`

The Uniswap V3 Factory.
A factory for creating new V3 trading pairs.

Creates new pairs at deterministic addresses.

## `allPairsLength() → uint256` (external)

Gets length of allPairs array.

: length of allPairs address array.

## `allEnabledFeeOptionsLength() → uint256` (external)

Gets length of allEnabledFeeOptions array.

: of allEnabledFeeOptions array.

## `constructor(address _owner)` (public)

The Factory contract constructor.

\_owner: The owner of the Factory contract.

## `createPair(address tokenA, address tokenB, uint24 fee) → address pair` (external)

Deploys a new trading pair.

tokenA: the first token of the desired pair.

tokenB: the second token of the desired pair.

fee: the desired fee.

: Returns the address of the newly deployed pair.

## `setOwner(address _owner)` (external)

Sets Factory contract owner to a new address.

only callable by current owner of factory contract.

\_owner: The new owner of the factory contract.

## `enableFeeOption(uint24 fee)` (external)

If chosen, enables the fee option when a pair is deployed.

fee: The chosen fee option - passed via createPair.
