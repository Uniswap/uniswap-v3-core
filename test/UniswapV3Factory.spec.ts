import {ethers, waffle} from 'hardhat'
import {UniswapV3Factory} from '../typechain/UniswapV3Factory'
import {expect} from './shared/expect'
import snapshotGasCost from './shared/snapshotGasCost'

import {FeeVote, getCreate2Address} from './shared/utilities'

const TEST_ADDRESSES: [string, string] = [
  '0x1000000000000000000000000000000000000000',
  '0x2000000000000000000000000000000000000000',
]

describe('UniswapV3Factory', () => {
  const [wallet, other] = waffle.provider.getWallets()

  let factory: UniswapV3Factory
  let pairBytecode: string
  beforeEach('deploy factory', async () => {
    const factoryFactory = await ethers.getContractFactory('UniswapV3Factory')
    pairBytecode = (await ethers.getContractFactory('UniswapV3Pair')).bytecode
    factory = (await factoryFactory.deploy(await wallet.getAddress())) as UniswapV3Factory
  })

  it('initial feeToSetter is deployer', async () => {
    expect(await factory.feeToSetter()).to.eq(wallet.address)
  })

  it('initial pairs length is 0', async () => {
    expect(await factory.allPairsLength()).to.eq(0)
  })

  async function createPair(tokens: [string, string], feeVote: FeeVote) {
    const create2Address = getCreate2Address(factory.address, tokens, feeVote, pairBytecode)
    await expect(factory.createPair(tokens[0], tokens[1], feeVote))
      .to.emit(factory, 'PairCreated')
      .withArgs(TEST_ADDRESSES[0], TEST_ADDRESSES[1], feeVote, create2Address, 1)

    await expect(factory.createPair(tokens[0], tokens[1], feeVote)).to.be.revertedWith(
      'UniswapV3::createPair: pair already exists'
    )
    await expect(factory.createPair(tokens[1], tokens[0], feeVote)).to.be.revertedWith(
      'UniswapV3::createPair: pair already exists'
    )
    expect(await factory.getPair(tokens[0], tokens[1], feeVote)).to.eq(create2Address)
    expect(await factory.getPair(tokens[1], tokens[0], feeVote)).to.eq(create2Address)
    expect(await factory.allPairs(0)).to.eq(create2Address)
    expect(await factory.allPairsLength()).to.eq(1)

    const pairContractFactory = await ethers.getContractFactory('UniswapV3Pair')
    const pair = pairContractFactory.attach(create2Address)
    expect(await pair.factory()).to.eq(factory.address)
    expect(await pair.token0()).to.eq(TEST_ADDRESSES[0])
    expect(await pair.token1()).to.eq(TEST_ADDRESSES[1])
    expect(await pair.fee()).to.eq(feeVote)
  }

  describe('#createPair', () => {
    it('succeeds', async () => {
      await createPair(TEST_ADDRESSES, FeeVote.FeeVote3)
    })

    it('succeeds in reverse', async () => {
      await createPair([TEST_ADDRESSES[1], TEST_ADDRESSES[0]], FeeVote.FeeVote1)
    })

    it('gas', async () => {
      await snapshotGasCost(factory.createPair(TEST_ADDRESSES[0], TEST_ADDRESSES[1], FeeVote.FeeVote0))
    })
  })

  describe('#setFeeToSetter', () => {
    it('fails if caller is not feeToSetter', async () => {
      await expect(factory.connect(other).setFeeToSetter(wallet.address)).to.be.revertedWith(
        'UniswapV3::setFeeToSetter: must be called by feeToSetter'
      )
    })

    it('updates feeToSetter', async () => {
      await factory.setFeeToSetter(other.address)
      expect(await factory.feeToSetter()).to.eq(other.address)
    })

    it('cannot be called by original feeToSetter', async () => {
      await factory.setFeeToSetter(other.address)
      await expect(factory.setFeeToSetter(wallet.address)).to.be.revertedWith(
        'UniswapV3::setFeeToSetter: must be called by feeToSetter'
      )
    })
  })
})
