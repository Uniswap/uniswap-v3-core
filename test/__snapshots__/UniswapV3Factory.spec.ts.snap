// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UniswapV3Factory #createPool gas 1`] = `4392554`;

exports[`UniswapV3Factory factory bytecode size 1`] = `23688`;

exports[`UniswapV3Factory pool bytecode size 1`] = `21309`;
